### ISSUE TITLE
Bug in the batch size vs accumulation step

### ISSUE THREAD

**Issue by PierreColombo** [axolotl-ai-cloud/axolotl#1100]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Hi ! 
Thanks for your great work. For the same effective batch, the loss and overall behavior should be the same.

When i do:
- Batch size 2, accumulation 8 i got decent loss (3.8)
- Batch size 16, accumulation 1, i got very high loss (over 8)

I do continue pertaining.

I guess there is something off in the way the loss / parameters are set.

### Current behaviour

To high loss

### Steps to reproduce

Make examples with varying combination of batch size accumulation. Maybe I'm missing something

### Which Operating Systems are you using?

- [ ] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main/4d6490b

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by RicardoDominguez** [comment]
Are you using sample packing and/or flash attention?

**Comment by PierreColombo** [comment]
No sample packing with Flash attention :)

**Comment by PierreColombo** [comment]
Same with this config:

`base_model: toto/toto
model_type: MistralForCausalLM
tokenizer_type: LlamaTokenizer
is_mistral_derived_model: true
resize_token_embeddings_to_32x: true

datasets:
  - path: data_processed/toto.jsonl
    ds_type: json
    type: sharegpt
    conversation: mistral

load_in_8bit: false
load_in_4bit: false
strict: false

dataset_prepared_path: toto
val_set_size: 0.005
output_dir: toto

sequence_len: 4096
sample_packing: false

dataset_shard_num: 1
# Index of shard to use for whole dataset
dataset_shard_idx: 0



wandb_mode: offline  # "offline" to save run metadata locally and not sync to the server, "disabled" to turn off wandb
wandb_project: toto # Your wandb project name
wandb_name: toto # Set the name of your wandb run
wandb_log_model: toto

gradient_accumulation_steps: 1
micro_batch_size: 12
num_epochs: 1
optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 1e-5
adam_beta1: 0.9
adam_beta2: 0.999
# Gradient clipping max norm
max_grad_norm: 1


warmup_steps:
evals_per_epoch: 5
eval_table_size:
saves_per_epoch: 5


group_by_length: false
# TODO :changer
bf16: true #true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true
max_steps:
debug:
deepspeed: deepspeed/zero3_bf16.json
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
`

**Comment by PierreColombo** [comment]
either i'm not doing something correctly or there is something off 👍  in the way accumulation and micro batch size influences the losses ?

Is this related: https://discuss.huggingface.co/t/gradient-accumulation-gives-different-results-compared-to-full-batch/65889 ?

**Comment by Palmik** [comment]
Is this train loss? Do you see effect on eval loss?
I could imagine that perhaps loss averaging is not handled consistently.

**Comment by PierreColombo** [comment]
The loss averaging is not handled consistantly but same with eval and training loss when batched !

**Comment by bursteratom** [comment]
Closing this as this is a known bug with gradient accumulation from transformers upstream which gives a higher loss when enabled. See here for most current unofficial fix: https://github.com/axolotl-ai-cloud/axolotl/pull/2064
